pkg/execution/debounce/debounce.go

```diff
@@ -0,0 +1,362 @@
+	FunctionVersion int `json:"fnV"`
+}
+
+// Debouncer represents an implementation-agnostic function debouncer, delaying function runs
+// until a specific time period passes when no more events matching a key are received.
+type Debouncer interface {
+	Debounce(ctx context.Context, d DebounceItem, fn inngest.Function) error
+	GetDebounceItem(ctx context.Context, debounceID ulid.ULID) (*DebounceItem, error)
+	DeleteDebounceItem(ctx context.Context, debounceID ulid.ULID) error
+}
+
+func NewRedisDebouncer(r rueidis.Client, k redis_state.DebounceKeyGenerator, q redis_state.QueueManager) Debouncer {
+	return debouncer{
+		r: r,
+		k: k,
+		q: q,
+	}
+}
+
+type debouncer struct {
+	r rueidis.Client
+	k redis_state.DebounceKeyGenerator
+	q redis_state.QueueManager
+}
+
+// DeleteDebounceItem returns a DebounceItem given a debounce ID.
+func (d debouncer) DeleteDebounceItem(ctx context.Context, debounceID ulid.ULID) error {
+	keyDbc := d.k.Debounce(ctx)
+	cmd := d.r.B().Hdel().Key(keyDbc).Field(debounceID.String()).Build()
+	err := d.r.Do(ctx, cmd).Error()
+	if rueidis.IsRedisNil(err) {
+		return nil
+	}
+	return fmt.Errorf("error removing debounce: %w", err)
+}
+
+// GetDebounceItem returns a DebounceItem given a debounce ID.
+func (d debouncer) GetDebounceItem(ctx context.Context, debounceID ulid.ULID) (*DebounceItem, error) {
+	keyDbc := d.k.Debounce(ctx)
+
+	cmd := d.r.B().Hget().Key(keyDbc).Field(debounceID.String()).Build()
+	byt, err := d.r.Do(ctx, cmd).AsBytes()
+	if rueidis.IsRedisNil(err) {
+		return nil, ErrDebounceNotFound
+	}
+
+	di := &DebounceItem{}
+	if err := json.Unmarshal(byt, &di); err != nil {
+		return nil, fmt.Errorf("error unmarshalling debounce item: %w", err)
+	}
+	return di, nil
+}
+
+// Debounce debounces a given function with the given DebounceItem.
+func (d debouncer) Debounce(ctx context.Context, di DebounceItem, fn inngest.Function) error {
+	if fn.Debounce == nil {
+		return fmt.Errorf("fn has no debounce config")
+	}
+	ttl, err := str2duration.ParseDuration(fn.Debounce.Period)
+	if err != nil {
+		return fmt.Errorf("invalid debounce duration: %w", err)
+	}
+	return d.debounce(ctx, di, fn, ttl, 0)
+}
+
+func (d debouncer) debounce(ctx context.Context, di DebounceItem, fn inngest.Function, ttl time.Duration, n int) error {
+	// Call new debounce immediately.  If this returns ErrDebounceExists then
+	// update the debounce.  This ensures that checking and creating a debounce
+	// is atomic, and two individual threads/workers cannot create debounces simultaneously.
+	debounceID, err := d.newDebounce(ctx, di, fn, ttl)
+	if err == nil {
+		return nil
+	}
+	if err != ErrDebounceExists {
+		// There was an unkown error creating the debounce.
+		return err
+	}
+	if debounceID == nil {
+		return fmt.Errorf("expected debounce ID when debounce exists")
+	}
+
+	// A debounce must already exist for this fn.  Update it.
+	err = d.updateDebounce(ctx, di, fn, ttl, *debounceID)
```

Comment:
are there potential race conditions here since they're separate calls?

Reply:
Nope.  Updating a debounce, if a debounce doesn't exist, will fail, re-invoking this function.  Debounce minimums are 1 second, so we're okay.